clean/post_officer_history.py

```diff
@@ -0,0 +1,272 @@
+                )
+                .str.replace(
+                    r"^Housing Authorityno$",
+                    "Housing Authority of New Orleans",
+                    regex=True,
+                )
+                .str.replace(r"^Univ PDtulane$", "Tulane University PD", regex=True)
+                .str.replace(r"Univ PDuno", "UNO University PD")
+                .str.replace(r"\bLa\b", "Louisiana")
+                .str.replace(r"^PlaqueminesSO$", "Plaquemines SO", regex=True)
+                .str.replace(r"\bNo\b", "New Orleans PD", regex=True)
+                .str.replace(r"^St\. Martin$", "St. Martin SO")
+                .str.replace(r"^Tangipahoa$", "Tangipahoa SO", regex=True)
+                .str.replace(r"^Lake Charles$", "Lake Charles PD")
+                .str.replace(r"^St\. Tammany$", "St. Tammany SO", regex=True)
+                .str.replace(r"^Hammond$", "Hammond PD", regex=True)
+                .str.replace(r"^New Orleans$", "New Orleans PD")
+                .str.replace(r"^Lafayette City$", "Lafayette City PD", regex=True)
+                .str.replace(r"^Grand Isle$", "Grand Isle PD", regex=True)
+                .str.replace(r"^New Orleans Levee$", "New Orleans Levee PD", regex=True)
+                .str.replace(r"^Harbor PD$", "New Orleans Harbor PD", regex=True)
+                .str.replace(r"^Lafayette$", "Lafayette PD", regex=True)
+                .str.replace(r"^Louisiana State$", "Louisiana State PD", regex=True)
+                .str.replace(r"^Caddo Parish$", "Caddo SO", regex=True)
+            )
+    return df
+
+
+def drop_duplicates(df):
+    return df.drop_duplicates(subset=["first_name", "last_name"], keep="last")
+
+
+def generate_history_id(df):
+    stacked_agency_sr = df[
+        [
+            "agency",
+            "agency_1",
+            "agency_2",
+            "agency_3",
+            "agency_4",
+            "agency_5",
+            "agency_6",
+            "agency_7",
+            "agency_8",
+        ]
+    ].stack()
+
+    stacked_agency_df = stacked_agency_sr.reset_index().iloc[:, [0, 2]]
+    stacked_agency_df.columns = ["history_id", "agency"]
+
+    names_df = df[["first_name", "last_name", "middle_name"]].reset_index()
+    names_df = names_df.rename(columns={"index": "history_id"})
+
+    stacked_agency_df = stacked_agency_df.merge(names_df, on="history_id", how="right")
+
+    return stacked_agency_df
+
+
+def check_for_duplicate_uids(df):
+    uids = df.groupby(["uid"])["history_id"].agg(list).reset_index()
+
+    for row in uids["history_id"]:
+        unique = all(element == row[0] for element in row)
+        if unique:
+            continue
+        else:
+            raise ValueError("uid found in multiple history ids")
+
+    return df[~((df.agency.fillna("") == ""))]
+
+
+def generate_false_neg_id(df):
+    false_negs = [
+        ids == False for ids in df.duplicated(subset=["history_id"], keep=False)
+    ]
+    df.loc[:, "false_negatives"] = false_negs
```

Comment:
Don't convert a series into a list and even discard elements from it. `df.duplicated(subset=["history_id"], keep=False)` is a boolean series which have the same index and the same length as `df`, so it is perfect to assign to a column already. `false_negs` not only  does not have any index (because it was converted into a list), it doesn't even have the same length. So I'm very curious how this assignment is even possible.

Reply:
Thanks.